File: 2025/day3/zig/day3.zig
// Copyright (c) 2025 Sabarish. All Rights Reserved.
const std = @import("std");

fn swapper(comptime N: usize, characters: *[N]u8, sv: []const u8, check_id: usize, char_id: usize) void {
    const elem = sv[char_id];
    var reset: bool = false;
    for (check_id..N) |j| {
        const elem2 = characters[j];
        if (elem > elem2 and !reset) {
            characters[j] = elem;
            reset = true;
            continue;
        }
        if (reset) {
            characters[j] = '0';
        }
    }
}
fn problem(comptime N: u32, lines: []const u8, res: *i64) !void {
    var split_lines = std.mem.splitScalar(u8, lines, '\n');
    while (split_lines.next()) |line| {
        var characters = [_]u8{'0'} ** N;
        var it: usize = 0;
        while (it < line.len) {
            const rem = line.len - it;
            const check_from: usize = if (rem > N) 0 else N - rem;
            swapper(N, &characters, line, check_from, it);
            it += 1;
        }
        var val: i64 = 0;
        for (characters) |elem| {
            const num = elem - '0';
            val = 10 * val + num;
        }
        res.* += val;
    }
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    const allocator = gpa.allocator();

    var args = try std.process.argsWithAllocator(allocator);
    defer args.deinit();
    _ = args.skip();

    const filepath = args.next() orelse {
        std.debug.print("No filepath provided\n", .{});
        return error.MissingArgument;
    };
    std.fs.cwd().access(filepath, .{}) catch |err| {
        std.debug.print("File {} does not exist!\n", .{err});
        return;
    };
    const file = try std.fs.cwd().openFile(filepath, .{});
    const contents = try file.readToEndAlloc(allocator, 20500);
    defer allocator.free(contents);
    var res: i64 = 0;
    try problem(2, contents, &res);
    std.debug.print("{d}\n", .{res});
    var res2: i64 = 0;
    try problem(12, contents, &res2);
    std.debug.print("{d}", .{res2});
}

test "example" {
    var res: i64 = 0;
    const data =
        "987654321111111\n811111111111119\n234234234234278\n818181911112111";
    try problem(2, data, &res);
    std.debug.assert(357 == res);
}

File: 2025/day3/go/day3.go
package main

import (
	"fmt"
	"os"
)

func swapper(N int, characters []byte, sv []byte, checkID int, charID int) {
	elem := sv[charID]
	reset := false
	for j := checkID; j < N; j++ {
		elem2 := characters[j]
		if elem > elem2 && !reset {
			characters[j] = elem
			reset = true
			continue
		}
		if reset {
			characters[j] = '0'
		}
	}
}

func problem(N int, lines []byte, res *int64) error {
	start := 0
	for start < len(lines) {
		// find next newline
		end := start
		for end < len(lines) && lines[end] != '\n' {
			end++
		}
		line := lines[start:end]
		start = end + 1

		if len(line) == 0 {
			continue
		}

		// characters = array filled with '0'
		characters := make([]byte, N)
		for i := range N {
			characters[i] = '0'
		}

		it := 0
		for it < len(line) {
			rem := len(line) - it
			checkFrom := 0
			if rem <= N {
				checkFrom = N - rem
			}
			swapper(N, characters, line, checkFrom, it)
			it++
		}

		// accumulate value
		var val int64 = 0
		for _, elem := range characters {
			num := int64(elem - '0')
			val = 10*val + num
		}
		*res += val
	}
	return nil
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("No filepath provided")
		return
	}
	filepath := os.Args[1]

	_, err := os.Stat(filepath)
	if err != nil {
		fmt.Printf("File does not exist: %v\n", err)
		return
	}

	data, err := os.ReadFile(filepath)
	if err != nil {
		fmt.Println("Error reading file:", err)
		return
	}

	var res int64 = 0
	problem(2, data, &res)
	fmt.Printf("%d\n", res)

	var res2 int64 = 0
	problem(12, data, &res2)
	fmt.Printf("%d\n", res2)
}
